### PR TITLE
Appveyor overhaul

### PR DESCRIPTION
**Pull request by janschulz**
Mostly:
* use a slightly simpler appveyor script ...
* enables conda package builds again by using a workaround for the libpng
* syncs the conda-recipe with conda-forge, including tightening the versioned dependencies to the one in conda-forge (we have to because we otherwise would break when someone installed libpng from defaults :-/)
* use a better install line in the conda-recipe build scripts (which is the recommended default for setuptool based ones
* revert the wheel workaround on windows -> newer wheels work without this...
* check that the wheels really have no external dll dependencies
* install more optional dependencies so that more tests are run (cc: https://github.com/matplotlib/matplotlib/issues/6523)

### CODE REVIEW

**Comment by jenshnielsen** [comment]
Looks good, Any chance we can drop the patched bdist_whell now?

**Comment by janschulz** [comment]
Yes, that can also be gone now. Will add it here in this PR

**Comment by janschulz** [comment]
Meh:

[matplotlib-1.5.0+2326.gf349629-cp27-cp27m-win_amd64.whl]
```
λ dumpbin.exe /DEPENDENTS /NOLOGO _png.pyd

Dump of file _png.pyd

File Type: DLL

  Image has the following dependencies:

    zlib.dll
    MSVCP90.dll
    python27.dll
    MSVCR90.dll
    KERNEL32.dll
```

-> depends on `zlib.dll`, so probably doesn't work at the user if zlib isn't inPATH for some other reason :-(

**Comment by janschulz** [comment]
Ok, it builds and the tests pass but the coverage is going down :-(

-> I would love to get a review, I don't have any more changes staged for this.

**Comment by jenshnielsen** [comment]
Don't worry about the coverage. It's comming from the travis builds and have been going up and down for that file all the time in various builds

**Comment by janschulz** [comment]
Ok, after installing inkscape and imagemagick the tests error out:

```
python tests.py
......KK.K.K.KK....SS...SSK.......SS...SSK.SS.SS..SS.....SSKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK.SS.SS.SS.SS...SS....SS.SS.SS.SS.SS.SS.SS.SS...SS.K.SS.SS.SS.SS.SS.....SS.SS.SS..SS.SS.SS.SS.SS.SS.SS.SS.SS.......SS........SS.SS.SS.SS......SS...SS.SS.SS.SS.SS.SS..SS.SS.SS...SS...........SS.SS.SS.SS.SS.SS.SS.SS.SS.SS.....SS...SS..SS.SS.SS.SS.............SS.SSc:\projects\matplotlib\lib\matplotlib\stackplot.py:95: RuntimeWarning: divide by zero encountered in true_divide
  inv_total = np.where(total > 0, 1./total, 0)
.SS.SS.SS.SS.SS.SS.SS....SS....................................................SSS........................SS....SSetProcessDpiAwareness failed: "COM error 0x80070005  (Unknown error 0x0ffffffff80070005)"
```

The test runtime of almost exactly 1h seem to suggest that the tests are stuck somewhere and run into a timeout :-( -> I'm removing the choko installs again...

**Comment by janschulz** [comment]
Ok, wasn't the choco lines, next try pillow... Please don't let it be miktex...

**Comment by cgohlke** [comment]
Re possible MiKTeX timeouts: check that MiKTeX is configured to "install missing packages on-the-fly", not "ask me first" (the default; opens a modal dialog).

**Comment by janschulz** [comment]
@cgohlke: Thank you, the miktex "ask" setting seems to have been the problem!

* old: OK (KNOWNFAIL=75, SKIP=1214)
* new without miktex: (KNOWNFAIL=82, SKIP=1214)
* new with miktex: OK (KNOWNFAIL=83, SKIP=1213)

-> It seems it still misses most of the tests due to missing tools :-(

**Comment by jenshnielsen** [comment]
The images are converted using inkscape but I guess choko installing inkscape does not put it on path so it can be found by nose

**Comment by janschulz** [comment]
Current status: A failure on py3.4: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1712

```
======================================================================
FAIL: matplotlib.tests.test_backend_ps.test_savefig_to_stringio_eps_afm
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Miniconda3-x64\envs\test-environment\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\projects\matplotlib\lib\matplotlib\testing\decorators.py", line 152, in wrapped_callable
    func(*args, **kwargs)
  File "c:\projects\matplotlib\lib\matplotlib\tests\test_backend_ps.py", line 89, in test_savefig_to_stringio_eps_afm
    _test_savefig_to_stringio(format='eps', use_log=True)
  File "c:\projects\matplotlib\lib\matplotlib\tests\test_backend_ps.py", line 54, in _test_savefig_to_stringio
    assert values[0] == values[1]
AssertionError
 
----------------------------------------------------------------------
Ran 6123 tests in 587.735s
```

Next stop: trying to put inkscape into PATH thanks to the suggestion from @jenshnielsen

**Comment by jenshnielsen** [comment]
@janschulz I think that one is a random fluke. We are seeing it on Travis too

**Comment by janschulz** [comment]
Still something missing:

```
Ran 6123 tests in 565.587s 
 
OK (KNOWNFAIL=84, SKIP=1213)
```

I think I know why: we only install bat files in PATH which redirect to the exe in a subdir for most of the files (only a few get a wrapper written in GO -> it's 2MB and we have a lot of binaries in the miktex package, so only the most mportant binaries got that treatment) and I think subprocess can't call bat files without using a shell which it doesn't in the checks: https://github.com/matplotlib/matplotlib/blob/98a2e64338664fd8b000007b3843caae81837439/lib/matplotlib/__init__.py#L355

**Comment by janschulz** [comment]
proper exe wrappers are comming: https://github.com/conda-forge/miktex-feedstock/pull/8

**Comment by janschulz** [comment]
Current status:

A lot less skipped tests: `(KNOWNFAIL=83, SKIP=609, failures=1)`. Tests now take 30min each instead of ~12 min :-(

But also a failure  on py27, py33

https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1721/job/4jt29336mqp6jx6g/artifacts
```
====================================================================== 
FAIL: matplotlib.tests.test_patches.test_wedge_range.test 
---------------------------------------------------------------------- 
Traceback (most recent call last):
  File "C:\Miniconda\envs\test-environment\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "c:\projects\matplotlib\lib\matplotlib\testing\decorators.py", line 55, in failer
    result = f(*args, **kwargs)
  File "c:\projects\matplotlib\lib\matplotlib\testing\decorators.py", line 267, in do_test
    '(RMS %(rms).3f)'%err)
ImageComparisonFailure: images not close: C:\projects\matplotlib\result_images\test_patches\wedge_range_pdf.png vs. C:\projects\matplotlib\result_images\test_patches\wedge_range-expected_pdf.png (RMS 1.617) 
```

Looking at the images (pdf versions, see the artifact in one of the failing appveyor runs), it needs a slightly higher tolerance?

**Comment by jenshnielsen** [comment]
I looked at the failed image. The diff seems insignificant so I am 👍  on changing the tolerance slightly

**Comment by janschulz** [comment]
ok, upped the tolerance on that test...

What is with the testing timeS: 4*30min=2h instead of 40min as before... I could install inkscape only on  one of the tests, so that we get back to 1x30+3*15min... Or is 2h ok?

**Comment by jenshnielsen** [comment]
I think installing inkscape  on only one of the nodes is probably the best solution at the moment.

**Comment by janschulz** [comment]
Any idea what's happening [here (Travis -> linux)](https://travis-ci.org/matplotlib/matplotlib/jobs/136425513#L1012):

```
 ======================================================================
FAIL: matplotlib.tests.test_backend_pdf.test_grayscale_alpha.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 55, in failer
    result = f(*args, **kwargs)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 257, in do_test
    self._tol, in_decorator=True)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/compare.py", line 337, in compare_images
    rms = calculate_rms(expectedImage, actualImage)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/compare.py", line 248, in calculate_rms
    "actual size {1}".format(expectedImage.shape, actualImage.shape))
matplotlib.testing.exceptions.ImageComparisonFailure: image sizes do not match expected size: (432, 576, 3) actual size (346, 461, 3)
```

**Comment by jenshnielsen** [comment]
It seems to happen randomly on both Travis and appveyor. It's a fluke not related to any on this but we should probably try to get to the bottom of this.

**Comment by janschulz** [comment]
It seems the choco install if inkscape is cached, at least it's now available in in all test runs :-/

**Comment by janschulz** [comment]
```
set LIBRARY_LIB=%CONDA_DEFAULT_ENV%\Library\lib
mkdir lib || cmd /c "exit /b 0"
A subdirectory or file lib already exists.
copy %LIBRARY_LIB%\zlibstatic.lib lib\z.lib
The system cannot find the path specified.
Command exited with code 1
```

WTF? I installed the save zlib in a local env and it worked...?

**Comment by janschulz** [comment]
Ok, the problem was that conda changed the `CONDA_DEFAULT_ENV` variable from containing the full path to the env to the env name only. This meant that paths which were based on that env variable were of course failing... The new way seems to be `CONDA_PREFIX`...

**Comment by janschulz** [comment]
I opened https://github.com/conda/conda-build/issues/1061 for the new "cannot find cl.exe" problem. 

Using conda and conda-build isn't the easiest thing to do right now :-(

File: appveyor.yml [hunk 1]
```diff
@@ -17,28 +17,33 @@ environment:
...
       CONDA_PY: "27"
       CONDA_NPY: "18"
       PYTHON_VERSION: "2.7"
+      TEST_ALL_IMAGES: "no"
       CONDA_INSTALL_LOCN: "C:\\Miniconda"
     - TARGET_ARCH: "x64"
       CONDA_PY: "27"
       CONDA_NPY: "18"
       PYTHON_VERSION: "2.7"
+      TEST_ALL_IMAGES: "yes" # only once, pick the one which made the most problems
```
Comment (QuLogic) [inline appveyor.yml]: Isn't 32-bit more likely to have problems?

Comment (janschulz) [inline appveyor.yml]: not sure. That was the one I had to do fixes for in this iteration. The last time it was a different version. I think that's probably just random?

**Comment by janschulz** [comment]
There is a new conda version out, lets see if this fixes things (close and reopen)

**Comment by tacaswell** [comment]
@janschulz you seem to be fighting windows, multi-process, setuptools, and conda* simultaneously right now via appevyor.

If we ever meet in person I owe you :beers: .

**Comment by janschulz** [comment]
Now it's this in only one of the runs:

```
C:\Miniconda-x64\conda-bld\work>call "C:\Program Files (x86)\Microsoft Visual Studio 9.0\VC\vcvarsall.bat" amd64 
Setting environment for using Microsoft Visual Studio 2008 x64 tools.
The input line is too long.
The syntax of the command is incorrect.
Command failed: C:\windows\system32\cmd.exe /c bld.bat
Command exited with code 1
```

**Comment by Kojoley** [comment]
May be this link will be helpful https://support.microsoft.com/en-us/kb/830473

> The input line is too long.
> The syntax of the command is incorrect.

**Comment by tacaswell** [comment]
Restarted the build to see if this is a real problem or something being flaky

**Comment by janschulz** [comment]
I'm starting to think that we should just remove the conda build again as long as it seems not reliable... :-(

**Comment by janschulz** [comment]
I've disabled the conda-build and removed the debugging output (will add a new PR for both...), so hopefully this succeeds now and can be merged...

**Comment by WeatherGod** [comment]
The tests are now all successful. Who should be the one to give the final review and check?

File: appveyor.yml (hunk 1 above)
Comment (QuLogic) [inline appveyor.yml]: I think maybe @cgohlke or @matthew-brett said 32-bit required the most finagling to build packages.

File: ci/conda_recipe/condaversion.patch [hunk 2]
```diff
@@ -11,4 +11,4 @@ index 8af8b6d..4e4f9d2 100644
 +    f.write(__version__)
  
  # These are the packages in the order we want to display them.  This
- # list may contain strings to create section headers for the display.
+ # list may contain strings to create section headers for the display.
```
Comment (QuLogic) [inline ci/conda_recipe/condaversion.patch]: Don't think this needs to be changed?

File: appveyor.yml (hunk 1 above)
Comment (janschulz) [inline appveyor.yml]: note that these controls *only* the image comparison tests for image formats *other than png*, png image tests (and all the rest of the tests) are run on all platforms.

File: ci/conda_recipe/condaversion.patch (hunk 2 above)
Comment (janschulz) [inline ci/conda_recipe/condaversion.patch]: yep, will change

File: appveyor.yml (hunk 1 above)
Comment (janschulz) [inline appveyor.yml]: changed to py27/32bit

**Comment by tacaswell** [comment]
Do we version pin something to make sure we are using a fixed version of the wheels build code?

**Comment by janschulz** [comment]
@tacaswell No, they are basically build against the same code which is in conda-forge at this time.

**Comment by tacaswell** [comment]
I am a bit concerned about removing the bug workaround in setup.py which may bite users on windows that want to build wheels which have not updated.

I am happy with saying that people must have newer versions of setuptools (I assume that is where it comes from?), but should probably put in a check for that.

I would also be happy to be told I am confused about this and it is fine as-is.

**Comment by janschulz** [comment]
I saw basically two bugreports about the wheel issue: mine and one from cgohlke who also tried compiling mpl. So IMO this is pretty rare: noone in the right mind tries to compile mpl from source to a wheel, everyone just gets it form cgohlke :-)

Wheel is from it's own package (at least under conda) but gets probably installed per default (at least conda installs it automatically in each new python env)?

**Comment by tacaswell** [comment]
Fair enough.

**Comment by janschulz** [comment]
Yay :-) Thanks!

**Comment by tacaswell** [comment]
Thank you for fighting the good fight with appveyor!

**Comment by Kojoley** [comment]
Build times have increased from ~10 min ([build <= 1.0.1921](https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1921)) to ~30 min ([build >= 1.0.1922](https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1922)) after merging this.

**Comment by janschulz** [comment]
before:
> Ran 6124 tests in 423.671s
> OK (KNOWNFAIL=74, SKIP=1214)

after:
> Ran 6125 tests in 1028.213s
> OK (KNOWNFAIL=82, SKIP=609)

And 64bit and the lone 32 bit (which was instructed to run all image comp tests) have the same runtime and the same number of tests.

Should I remove the additional requirements on all but one build, so only one gets down to 609 skips vs 1200 in the old days?

**Comment by Kojoley** [comment]
On my machine `inkscape` is terribly slow, I hope some day `librsvg` would replace it.

**Comment by tacaswell** [comment]
librsvg has other issues like not implementing the svg spec correctly (https://bugzilla.gnome.org/show_bug.cgi?id=748565).

**Comment by jenshnielsen** [comment]
I think @mdboom mentioned that they had fixed that issue but probably not in a release yet

**Pull request merged**